### PR TITLE
Add Sonobi Adapter

### PR DESCRIPTION
**Pull request by Studnicky**
Adapter code for Sonobi clients

### CODE REVIEW

File: src/adapters/sonobi.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
+var bidfactory = require('../bidfactory.js');
+var bidmanager = require('../bidmanager.js');
+var utils = require('../utils');
+
+var SonobiAdapter = function SonobiAdapter(){
+  var trinity_ep = 'http://apex.go.sonobi.com/trinity.js?key_maker=';
```
Comment (mkendall07) [inline src/adapters/sonobi.js]: Do you have a `HTTPS` endpoint? We are requiring all adapters to use secure tags now.

File: src/adapters/sonobi.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  var test = false; //  tag tester = true || false
+  var cb_map = {};
+
+  function _phone_in(params){
+    var bids = params.bids || [];
+    _insert_header(_keymaker(bids));
+  }
+
+  function _insert_header(keyring){ //  Inject trinity into header
```
Comment (mkendall07) [inline src/adapters/sonobi.js]: Please use `adloader.loadScript` instead
https://github.com/prebid/Prebid.js/blob/master/src/adloader.js#L5

File: src/adapters/sonobi.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      } else {          //  No aid? No ad.
+        bidObject = bidfactory.createBid(2);
+        bidObject.bidderCode = 'sonobi';
+        bidmanager.addBidResponse(cb_map[slot], bidObject);
+      }
+    }
+  }
+
+  function _get_creative(sbi_dc, sbi_aid){
+    var creative = '<scr' + 'ipt type="text/javascript"src="http://' + sbi_dc;
```
Comment (mkendall07) [inline src/adapters/sonobi.js]: Need `HTTPS` input should be encoded for safety against XSS.

File: src/adapters/sonobi.js (hunk 2 above)
Comment (Studnicky) [inline src/adapters/sonobi.js]: JSONP callbacks?

Comment (mkendall07) [inline src/adapters/sonobi.js]: Usage is `adloader.loadScript('scriptUrl', cb, cacheUrl)` 
For JSONP you can just append your callback to `scriptUrl`

File: src/adapters/sonobi.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+var utils = require('../utils');
+
+var SonobiAdapter = function SonobiAdapter(){
+  var test = false;               //  tag tester = true || false
+  var cb_map = {};
+
+  function _phone_in(params){
+    var trinity = '//apex.go.sonobi.com/trinity.js?key_maker=';
+    var bids = params.bids || [];
+    adloader.loadScript(trinity + JSON.stringify(_keymaker(bids)) + '&cv=' + _operator(), null);
```
Comment (mkendall07) [inline src/adapters/sonobi.js]: Thanks for updating. You don't need to explicitly pass `null` here

**Comment by mkendall07** [comment]
@Studnicky 
Could you please add your bidder in the example file `https://github.com/prebid/Prebid.js/blob/master/integrationExamples/gpt/pbjs_example_gpt.html` and also give us params validate a bid response (can be a test ad)? Thanks

**Comment by Studnicky** [comment]
Note that actual bidding still requires an account with Sonobi.

**Comment by Studnicky** [comment]
First commit will show you working test ads with our placements.
Last commit shows parameters in same format as other publishers.

File: src/adapters/sonobi.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+var bidmanager = require('../bidmanager.js');
+var adloader = require('../adloader.js');
+var utils = require('../utils');
+
+var SonobiAdapter = function SonobiAdapter(){
+  var test = false;               //  tag tester = true || false
+  var cb_map = {};
+
+  function _phone_in(params){
+    var trinity = '//apex.go.sonobi.com/trinity.js?key_maker=';
```
Comment (protonate) [inline src/adapters/sonobi.js]: We are now always using https in place of relative protocol urls.

File: src/adapters/sonobi.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    var bids = params.bids || [];
+    adloader.loadScript(trinity + JSON.stringify(_keymaker(bids)) + '&cv=' + _operator(), null);
+  }
+
+  function _keymaker(bids){       //  Make keys
+    var keyring = {};
+    bids.forEach(function(o){
```
Comment (protonate) [inline src/adapters/sonobi.js]: please use `utils._each(bids, function() { ... })` as we target support to IE8+.

File: src/adapters/sonobi.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      if (slots[slot].sbi_aid){
+        bidObject = bidfactory.createBid(1);
+        bidObject.bidderCode = 'sonobi';
+        bidObject.cpm = Number(slots[slot].sbi_mouse);
+        bidObject.ad = _get_creative(sbi_dc, slots[slot].sbi_aid);
+        bidObject.width = Number(slots[slot].sbi_size.split('x')[0]);
+        bidObject.height = Number(slots[slot].sbi_size.split('x')[1]);
+        bidmanager.addBidResponse(cb_map[slot], bidObject);
+      } else {                      //  No aid? No ad.
+        bidObject = bidfactory.createBid(2);
```
Comment (protonate) [inline src/adapters/sonobi.js]: I am getting a response when configuring using the test slots you referenced, and a placeholder creative is rendered to the slot with text e.g. "This is a 300 x 600" shown in the iframe, however `sbi_aid` is not in the response so the bid is created here as "Bid returned empty or error response".

Comment (Studnicky) [inline src/adapters/sonobi.js]: Yes, without using live bidding or a test-cookie to force a k:v return, that is the expected result.

Comment (Studnicky) [inline src/adapters/sonobi.js]: Here's a placement id that will return:

`{placement_id: "21d2da738fe0ba795cfb-test"}`

Comment (protonate) [inline src/adapters/sonobi.js]: Great, thanks, bids with creative returned with correct status.

**Comment by protonate** [comment]
Reviewed, please squash commits and we can merge.

**Comment by Studnicky** [comment]
Fixed history, merged down.

File: src/adapters/sonobi.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+  function _phone_in(params){
+    var trinity = 'https://apex.go.sonobi.com/trinity.js?key_maker=';
+    var bids = params.bids || [];
+    adloader.loadScript(trinity + JSON.stringify(_keymaker(bids)) + '&cv=' + _operator(), null);
+  }
+
+  function _keymaker(bids){       //  Make keys
+    var keyring = {};
+    bids._each(function(o){
```
Comment (protonate) [inline src/adapters/sonobi.js]: There is a bug here that throws an error as `._each` is not a function. You can use `utils._each(bids, function(o) { ... });

**Comment by protonate** [comment]
We need a single commit per PR as described in this [article](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) so as to keep the commit history a clear record of what was merged to master. I also see some unrelated commits in the history here so if you can open a new PR for this it would be easiest I think.

**Pull request closed without merging**